File: min/C++/1654.cpp
#include <iostream>
#include <vector>
using namespace std;

int main(void) {

	ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
	
	int k, n;
	long long num, max = -1;
	vector <int> v;

	cin >> k >> n;
	for (int i = 0; i < k; i++) {
		cin >> num;
		v.push_back(num);
		if (num > max) {
			max = num;
		}
	}

	long long int left = 1, mid, right = max, ans = -1;
	while (left <= right) {
		mid = (left + right) / 2;
		long long int temp = 0;

		for (int i = 0; i < k; i++) {
			temp += v[i] / mid;
		}

		if (temp >= n) {
			ans = mid;
			left = mid + 1;
		}
		else {
			right = mid - 1;
		}
	}

	cout << ans;

	return 0;
}

File: min/Python/1654.py
k, n = input().split()
k = int(k)
n = int(n)
arr = []
max = 0

for i in range(k):
    num = int(input())
    arr.append(num)
    if(num > max):
        max = num
    
left = 1
right = max
ans = 0

while(left<=right):
    mid = (left+right)//2
    sum = 0
    for i in range(k):
        sum += arr[i]//mid
    if(sum >= n):
        ans = mid
        left = mid+1
    else:
        right = mid-1

print(ans)
